File: server/src/intTest/java/org/tctalent/server/repository/db/GetCandidateSavedListsQueryIntTest.java
/*
 * Copyright (c) 2024 Talent Beyond Boundaries.
 *
 * This program is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License
 * for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see https://www.gnu.org/licenses/.
 */

package org.tctalent.server.repository.db;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.tctalent.server.repository.db.integrationhelp.DomainHelpers.getCandidateSavedList;
import static org.tctalent.server.repository.db.integrationhelp.DomainHelpers.getSavedCandidate;
import static org.tctalent.server.repository.db.integrationhelp.DomainHelpers.getSavedSavedList;
import static org.tctalent.server.repository.db.integrationhelp.DomainHelpers.getSavedUser;

import java.util.List;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.tctalent.server.model.db.Candidate;
import org.tctalent.server.model.db.CandidateSavedList;
import org.tctalent.server.model.db.SavedList;
import org.tctalent.server.repository.db.integrationhelp.BaseDBIntegrationTest;

public class GetCandidateSavedListsQueryIntTest extends BaseDBIntegrationTest {

  @Autowired
  private CandidateSavedListRepository repo;
  @Autowired
  private SavedListRepository savedListRepository;
  @Autowired
  private CandidateRepository candidateRepository;
  @Autowired
  private UserRepository userRepository;
  private SavedList testSavedList;
  private Candidate testCandidate;

  @BeforeEach
  public void setup() {
    assertTrue(isContainerInitialised());
    testCandidate = getSavedCandidate(candidateRepository, getSavedUser(userRepository));
    testSavedList = getSavedSavedList(savedListRepository);

    CandidateSavedList candidateSavedList = getCandidateSavedList(testCandidate, testSavedList);
    repo.save(candidateSavedList);
    assertNotNull(candidateSavedList.getId());
  }

  @Test
  public void testCandidateSavedListsQuery() {
    GetCandidateSavedListsQuery spec = new GetCandidateSavedListsQuery(testCandidate.getId());
    List<SavedList> result = savedListRepository.findAll(spec);
    assertNotNull(result);
    assertFalse(result.isEmpty());
    assertEquals(1, result.size());
    assertEquals(testSavedList.getId(), result.getFirst().getId());
  }

  @Test
  public void testCandidateSavedListsQueryWithNoMatchingCandidate() {
    GetCandidateSavedListsQuery spec = new GetCandidateSavedListsQuery(-1L);
    List<SavedList> result = savedListRepository.findAll(spec);
    assertNotNull(result);
    assertTrue(result.isEmpty());
  }
}

File: server/src/intTest/kotlin/org/tctalent/server/repository/db/GetCandidateSavedListsQueryIntTest.kt
/*
 * Copyright (c) 2024 Talent Beyond Boundaries.
 *
 * This program is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License
 * for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see https://www.gnu.org/licenses/.
 */

package org.tctalent.server.repository.db

import kotlin.test.*
import org.springframework.beans.factory.annotation.Autowired
import org.tctalent.server.model.db.Candidate
import org.tctalent.server.model.db.CandidateSavedList
import org.tctalent.server.model.db.SavedList
import org.tctalent.server.repository.db.integrationhelp.*

class GetCandidateSavedListsQueryIntTest : BaseDBIntegrationTest() {
  @Autowired lateinit var repo: CandidateSavedListRepository
  @Autowired lateinit var savedListRepository: SavedListRepository
  @Autowired lateinit var candidateRepository: CandidateRepository
  @Autowired lateinit var userRepository: UserRepository

  private lateinit var candidateSavedList: CandidateSavedList
  private lateinit var testSavedList: SavedList
  private lateinit var testCandidate: Candidate

  @BeforeTest
  fun setup() {
    assertTrue { isContainerInitialized() }
    testCandidate = getSavedCandidate(candidateRepository, getSavedUser(userRepository))
    testSavedList = getSavedSavedList(savedListRepository)

    candidateSavedList = getCandidateSavedList(testCandidate, testSavedList)
    repo.save(candidateSavedList)
    assertNotNull(candidateSavedList.id)
  }

  @Test
  fun `test candidate saved lists query`() {
    val spec = GetCandidateSavedListsQuery(testCandidate.id)
    val result = savedListRepository.findAll(spec)
    assertNotNull(result)
    assertTrue(result.isNotEmpty())
    assertEquals(1, result.size)
    assertEquals(testSavedList.id, result.first().id)
  }

  @Test
  fun `test candidate saved lists query with no matching candidate`() {
    val spec = GetCandidateSavedListsQuery(-1L)
    val result = savedListRepository.findAll(spec)
    assertNotNull(result)
    assertTrue(result.isEmpty())
  }
}
